### PR TITLE
add InstancePropertyHelper and apply_request_extensions

### PR DESCRIPTION
**Pull request by mmerickel**

### CODE REVIEW

**Comment by bertjwregeer** [comment]
LGTM

**Comment by mmerickel** [comment]
That little edit to IRequestFactory was because IRequestFactory is a public api and IRequest is apparently not. So I didn't want the private api referenced from the public docstring.

**Comment by bertjwregeer** [comment]
I had not realised that `IRequest` is not a public API. Probably because at this point it is excessively large ;-)

**Comment by mmerickel** [comment]
Well I actually thought it was. But it's private and just a marker interface actually.

File: pyramid/request.py [hunk 1]
```diff
@@ -307,3 +312,22 @@ def call_app_with_subpath_as_path_info(request, app):
...
     return new_request.get_response(app)
+
+def apply_request_extensions(request, extensions=None):
+    """Apply request extensions (methods and properties) to an instance of
+    :class:`pyramid.interfaces.IRequest`. This method is dependent on the
+    ``request`` containing a properly initialized registry.
+
+    After invoking this method, the ``request`` should have the methods
+    and properties that were defined using
+    :meth:`pyramid.config.Configurator.add_request_method`.
```
Comment (sontek) [inline pyramid/request.py]: Should we document how you would use `extensions`?

File: pyramid/router.py [hunk 2]
```diff
@@ -213,7 +214,7 @@ def invoke_subrequest(self, request, use_tweens=False):
             try:
                 extensions = self.request_extensions
                 if extensions is not None:
-                    request._set_extensions(extensions)
+                    apply_request_extensions(request, extensions=extensions)
```
Comment (sontek) [inline pyramid/router.py]: `self.request_extensions` is just the interface that `apply_request_extensions` queries anyways, do we need to pass them in?

Comment (mmerickel) [inline pyramid/router.py]: This is just an optimization to have the extensions queried from the registry only once at app-setup-time. This code is in pyramid's hot route, so I wanted to keep it as close to the original version as possible.

File: pyramid/request.py (hunk 1 above)
Comment (mmerickel) [inline pyramid/request.py]: I explicitly left `extensions` off of the public api documentation. It's there just for pyramid's hot route in the router.

Comment (mmerickel) [inline pyramid/request.py]: We could name it `_extensions` if it would make you feel better.

**Pull request merged**